[OmniLink] Binding Initial Contribution

ecdye opened the pull request:
Initial contribution / migration from OH1 of the OmniLink Binding. Ported with permission from @craigham and @boc-tothefuture who were the primary developers of this code.

This was created to add support to the main OH code-base for the OmniLink protocol to allow for more usage of it's abilities.

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

digitaldan commented:
Hi @ecdye to be clear, you are submitting the work that @boc-tothefuture, @craigham, and myself had worked on as part of the OH2/3 port that we never got across the line for review (my guess is there will be a lot of review issues). Are you volunteering to maintain this code going forward and update it as necessary, both when this gets reviewed and after it gets published?  Have you contributed any code to this binding yet other than some cosmetic logging changes?  If you are planning on maintaining and improving it that would be awesome, but if not i would want our reviewers to know as there is a lot of code to review.

digitaldan commented:
I also noticed you forked my jomnilink project , made a small change to use a newer java version, then repointed this binding to your own version instead of submitting a PR on my project?  any reason why? Would be nice to update the main one which gets pushed to maven central and used by other people.

ecdye commented:
> Hi @ecdye to be clear, you are submitting the work that @boc-tothefuture, @craigham, and myself had worked on as part of the OH2/3 port that we never got across the line for review (my guess is there will be a lot of review issues). 

Correct.

> Are you volunteering to maintain this code going forward and update it as necessary, both when this gets reviewed and after it gets published? 

Yes, for the foreseeable future.

> Have you contributed any code to this binding yet other than some cosmetic logging changes?  If you are planning on maintaining and improving it that would be awesome, but if not i would want our reviewers to know as there is a lot of code to review.

Nothing major, I have rewritten small parts for OH3 and fixed broken support for certain 'Things' like the Console Thing. I do not know how many new features I would add but I would certainly like to improve the current ones and make the binding more user friendly overall.

ecdye commented:
> I also noticed you forked my jomnilink project , made a small change to use a newer java version, then repointed this binding to your own version instead of submitting a PR on my project?  any reason why? Would be nice to update the main one which gets pushed to maven central and used by other people.  


I was simply testing a change that I intend to submit to your repo but haven't yet. The change was relating to the only outstanding issue in your jomnilink repo. I will make a PR for you to review and incorporate in the next few days.

boc-tothefuture commented:
@ecdye I think it is important to note @digitaldan contributions to this binding and the openhab library.  It wasn't just @craigham and myself..  I am also willing to support this binding going forward along with you.  I just didn't have the time to put into the documentation to get it ready.

digitaldan commented:
>Yes, for the foreseeable future.

Sounds great, happy to see someone pushing this across the line, although its a pity Leviton discontinued the Omni line of panels.  Awesome that @boc-tothefuture volunteered to help,  I'm available if you get stuck as well, although I just ordered a konnected.io system and plan to play around with that + noonlight for monitoring in leu of my omni, but thats a little ways off from being proved out, so my omni is still in use for the foreseeable future.

boc-tothefuture commented:
I would like to see @craigham sign off on this as well.. As it is his original work we all iterated off of.  I would also think that the three of us (@digitaldan, @craigham and myself) would need to officially sign this work since we wrote the majority of it?

craigham commented:
I have no problem handing this off to someone that can try and get this accepted.  Lately I have become more of a user of openhab rather than helping develop bindings.

I do have 1 question.  I have worked on several bindings in the past and it was nice to have each in a different branch within my addons fork.  This PR looks to be merging to main.  Is this necessary, or could it be done on a more specifically named branch?  

What do you think @boc-tothefuture, @digitaldan ?

boc-tothefuture commented:
> I have no problem handing this off to someone that can try and get this accepted. Lately I have become more of a user of openhab rather than helping develop bindings.
> 
> I do have 1 question. I have worked on several bindings in the past and it was nice to have each in a different branch within my addons fork. This PR looks to be merging to main. Is this necessary, or could it be done on a more specifically named branch?
> 
> What do you think @boc-tothefuture, @digitaldan ?

I think it goes to main when we PR it for official review.

ecdye commented:
> @ecdye I think it is important to note @digitaldan contributions to this binding and the openhab library. It wasn't just @craigham and myself.. I am also willing to support this binding going forward along with you. I just didn't have the time to put into the documentation to get it ready.

I agree, it would be important for @digitaldan to sign it off, I am merely the one who is facilitating the review and submission of this. 

>I do have 1 question. I have worked on several bindings in the past and it was nice to have each in a different branch within my addons fork. This PR looks to be merging to main. Is this necessary, or could it be done on a more specifically named branch?

The main branch is used to facilitate the central deployment and development of bindings, in my branch of the openhab-addons repo, it is being developed on a separate branch.

craigham commented:
ah, i see.  it's going getting submitted now to openhab for review.  great,
that would be nice.

If there is something I can help with, I will.  I've given up keeping an up
to date dev environment for openhab, but I could get it going again if i
can help.

Is there a formal way I need to sign off on any of the code?

Would be really nice to see the binding approved...



On Mon, 2 Nov 2020 at 09:10, boc-tothefuture <notifications@github.com>
wrote:

> I have no problem handing this off to someone that can try and get this
> accepted. Lately I have become more of a user of openhab rather than
> helping develop bindings.
>
> I do have 1 question. I have worked on several bindings in the past and it
> was nice to have each in a different branch within my addons fork. This PR
> looks to be merging to main. Is this necessary, or could it be done on a
> more specifically named branch?
>
> What do you think @boc-tothefuture <https://github.com/boc-tothefuture>,
> @digitaldan <https://github.com/digitaldan> ?
>
> I think it goes to main when we PR it for official review.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab-addons/pull/8922#issuecomment-720604465>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACXGUXQCRFCRCHHZY7L5CHDSN3RZ5ANCNFSM4TGIOUMQ>
> .
>

ecdye commented:
I have now linked everyone who helped in the creation of this binding as co-authors in the commit.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/OmnilinkDiscoveryService.java:
```diff
@@ -0,0 +1,543 @@
+     * @throws OmniUnknownMessageTypeException
+     * @throws BridgeOfflineException
+     */
+    private void discoverButtons()
+            throws OmniInvalidResponseException, OmniUnknownMessageTypeException, BridgeOfflineException {
```
These exceptions are never thrown. Same for several other method signatures in this class.

fwolter commented on bundles/org.openhab.binding.omnilink/README.md:
```diff
@@ -0,0 +1,218 @@
+
+This binding integrates the [OmniPro and Lumina](http://www.leviton.com/en/products/security-automation/automation-av-controllers/omni-security-systems) line of home automation systems. At Its core the OmniPro is a hardware board that provides security and access features. It connects to many other devices through serial ports or wired contacts and exposes them through a single TCP based API.
+
+## Supported Things
```
Can you add a table with the Channels? You could take a look at other bindings for some examples.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/OmnilinkDiscoveryService.java:
```diff
@@ -0,0 +1,543 @@
+                        case UNIT_TYPE_VIZIARF_LOAD:
+                        case UNIT_TYPE_COMPOSE:
+                            // let's prepend room name to unit name for label
+                            // TODO could make this configurable
```
Can you implement it or remove the TODO?

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractAreaHandler.java:
```diff
@@ -0,0 +1,205 @@
+        int mode = getMode(channelUID);
+        int areaNumber = getThingNumber();
+
+        if (!(command instanceof StringType)) {
+            throw new IllegalArgumentException(
+                    MessageFormat.format("Command {} is not valid for channel {}, only StringType are accepted",
+                            command, channelUID.getId()));
```
As receiving invalid types should be treated as a normal use case, it seems unreasonable to throw an unchecked exception. You could simply log it. Same for below.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractOmnilinkHandler.java:
```diff
@@ -0,0 +1,91 @@
+        return ((Number) getThing().getConfiguration().get(THING_PROPERTIES_AREA)).intValue();
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        super.bridgeStatusChanged(bridgeStatusInfo);
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.ONLINE);
+        }
+    }
```
Actually the framework takes care of the Thing status when the bridge is offline. Is this method really necessary?

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractOmnilinkStatusHandler.java:
```diff
@@ -0,0 +1,100 @@
+                handleStatus(status.get());
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                logger.debug("Received null staus update!");
+                updateStatus(ThingStatus.UNKNOWN);
```
In case of an error the Thing should be set to offline. The message you are logging could be used as the `ThingStatusDetailMessage`. The framework will log it, then.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AudioSourceHandler.java:
```diff
@@ -0,0 +1,160 @@
+@NonNullByDefault
+public class AudioSourceHandler extends AbstractOmnilinkHandler {
+    private final Logger logger = LoggerFactory.getLogger(AudioSourceHandler.class);
+    private static final long POLL_DELAY = 5; // 5 Second polling
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AudioSourceHandler.java:
```diff
@@ -0,0 +1,160 @@
+
+    private synchronized void cancelPolling() {
+        final ScheduledFuture<?> scheduledPolling = this.scheduledPolling;
+        if (scheduledPolling != null && !scheduledPolling.isDone()) {
```
No harm in canceling a task that is already cancelled.
```suggestion
        if (scheduledPolling != null) {
```

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AudioSourceHandler.java:
```diff
@@ -0,0 +1,160 @@
+        cancelPolling();
+        int sourceNumber = getThingNumber();
+        logger.debug("Scheduling polling for Audio Source {}", sourceNumber);
+        scheduledPolling = super.scheduler.scheduleWithFixedDelay(() -> pollAudioSource(), 0, POLL_DELAY,
```
Syntactical sugar.
```suggestion
        scheduledPolling = super.scheduler.scheduleWithFixedDelay(this::pollAudioSource, 0, POLL_DELAY,
```

fwolter reviewed (changes requested):
I reviewed 17 / 55 files that you get an impression what needs to be done.

Is the library jomnilink used by other projects than OH? How many active maintainers does it have?

ecdye commented:
@Fwolter Thanks for the suggestions, I have made many uncommitted improvements as well and will add these concerns into my changes. jomnilink is not used by any other OH addons however it is essential to communicating with the OmniPro or Lumina controller, it is maintained by primarily @digitaldan with minor contributions from myself and @boc-tothefuture. I currently have an outstanding PR on the jomnilink repo that will provide the solution for many of the uncomitted changes that I mentioned. 

Overall, once completed, with your suggestions, I feel that this PR will be much closer to a finished product.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/ObjectPropertyRequest.java:
```diff
@@ -0,0 +1,117 @@
+                    break;
+                }
+            } catch (OmniInvalidResponseException | OmniUnknownMessageTypeException | BridgeOfflineException e) {
+                logger.error("Error retrieving object properties", e);
```
Yep, been working on those in uncomitted changes.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/OmnilinkDiscoveryService.java:
```diff
@@ -0,0 +1,543 @@
+     * @param bridgeHandler Bride to discover things on
+     */
+    public OmnilinkDiscoveryService(OmnilinkBridgeHandler bridgeHandler) {
+        super(Collections.singleton(new ThingTypeUID(BINDING_ID, "-")), DISCOVER_TIMEOUT_SECONDS, false);
```
indicates all things, I'll update it

swamiller commented:
@ecdye I’m starting to dabble with OH3. If you need a tester, I’d be happy to help.

ecdye commented:
That would be great, mostly I going by what I know works because I use it. Right now I'm just waiting for @digitaldan to finish getting a PR for jomnilink so I can push a bunch of new changes.

randye007 commented:
Thanks so much @ecdye and @digitaldan for reviving the jomnilink binding for OH3. Is there a date you are working towards to getting this released?

Cheers,
   Randy

ecdye commented:
Not right now, still waiting on @digitaldan but he is busy, I don't see any need to rush.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmnilinkHandlerFactory.java:
```diff
@@ -0,0 +1,157 @@
+                discovery.unregister();
+            }
+        }
+    }
```
I would prefer to leave it as is for now as it works and reduces the complexity of my life for the moment, not having to worry about it.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/TemperatureFormat.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @author Craig Hamilton - Initial contribution
+ */
+public enum TemperatureFormat {
```
Is this still needed? CELSIUS doesn't seem to be used.

ecdye replied:
It is for cases where the system uses it.

fwolter replied:
Are both used FAHRENHEIT and CELSIUS?

ecdye replied:
Yes, depending on the system setup, if the controller is configured for Celsius, it will be used, if Fahrenheit it will be

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,194 @@
+        logger.debug("mode {} on zone {} with code {}", mode, zoneNumber, command.toFullString());
+        char[] code = command.toFullString().toCharArray();
+        if (code.length != 4) {
+            logger.error("Invalid code length, code must be 4 digits");
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be warn. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,194 @@
+                    logger.debug("Received null bridge while sending zone command!");
+                }
+            } catch (OmniInvalidResponseException e) {
+                logger.debug("Zone command failed", e);
+            } catch (OmniUnknownMessageTypeException | BridgeOfflineException e) {
+                logger.error("Could not send zone command", e);
```
See above. Logging stack trace and logging to error.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,194 @@
+                return Optional.empty();
+            }
+        } catch (OmniInvalidResponseException | OmniUnknownMessageTypeException | BridgeOfflineException e) {
+            logger.debug("Received exception while refreshing zone!", e);
```
See above. Logging stack trace.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>OmniLink Binding</name>
+	<description>This is the binding for OmniLink, a security system that interfaces with many devices.</description>
+	<author>Craig Hamilton</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,173 @@
+		<channels>
+			<channel id="sysdate" typeId="sysDate"/>
+			<channel id="enable_disable_beeper" typeId="console_enable_disable_beeper">
+				<label>Enable/Disable All Console Beepers</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,173 @@
+				<description>The IP or host name of the controller.</description>
+			</parameter>
+			<parameter name="port" type="integer" required="true">
+				<context>network-port</context>
```
Are you sure this context exists?

ecdye replied:
What do you mean?

fwolter replied:
`network-port` is not documented here https://www.openhab.org/docs/developer/bindings/config-xml.html#xml-structure-for-configuration-descriptions

I was wondering if you found an undocumented feature.

ecdye replied:
I actually don't know. It has been that way since i started developing for this binding. @digitaldan @craigham any idea?

craigham replied:
Not sure.  I believe @digitaldan may have written this.   It is just the ip port specification, however other approved bindings do this should be used.

digitaldan replied:
who knows, this was written long ago when we were moving from 1.x to 2.x and was probably a misunderstanding of the new XML format at the time. If you run into this type of thing its obviously just best to fix it since this clearly violates  the schema definition.

ecdye replied:
I'll remove it then

fwolter commented on bundles/org.openhab.binding.omnilink/src/main/resources/OH-INF/thing/unit.xml:
```diff
@@ -0,0 +1,332 @@
+
+	<channel-type id="on_for_seconds">
+		<item-type>Number</item-type>
+		<label>Turn Unit On For Seconds</label>
```
See above. Label length. Please also check the following channel-types.

fwolter reviewed (changes requested):
Your code doesn't compile.
`OmnilinkHandlerFactory.java:[37,8] The import org.openhab.binding.omnilink.internal.handler.units cannot be resolved` among others

boc-tothefuture commented on bundles/org.openhab.binding.omnilink/README.md:
```diff
@@ -0,0 +1,321 @@
+| status                      | Number               | The current numeric status of this thermostat/unit/dimmable unit.                    | thermostat, unit, dimmable          |
+| level                       | Dimmer               | Increase/Decrease the level of this unit/dimmable unit/UPB unit.                     | unit, dimmable, upb                 |
+| switch                      | Switch               | Turn this unit/dimmable unit/flag/output/room on/off.                                | unit, dimmable, flag, output, room  |
+| on_for_seconds              | Number               | Turn on this unit for a specified number of seconds.                                 | unit, dimmable, upb, flag, output   |
+| off_for_seconds             | Number               | Turn off this unit for a specified number of seconds.                                | unit, dimmable, upb, flag, output   |
+| on_for_minutes              | Number               | Turn on this unit for a specified number of minutes.                                 | unit, dimmable, upb, flag, output   |
+| off_for_minutes             | Number               | Turn off this unit for a specified number of minutes.                                | unit, dimmable, upb, flag, output   |
+| on_for_hours                | Number               | Turn on this unit for a specified number of hours.                                   | unit, dimmable, upb, flag, output   |
+| off_for_hours               | Number               | Turn off this unit for a specified number of hours.                                  | unit, dimmable, upb, flag, output   |
```
Maybe?  But it would be tricky. They are mapping directly to command of the same type on the backend system. 

So if you get a Number:Time of 90 seconds, you need to know to send it to seconds and not send 1 to minutes and then 30 seconds. 

And not time values are valid. 
So if you get a time of like 15 hours and 8 minutes and 9 seconds.  You can't just convert that to seconds because it would exceed the number of seconds you can send. So you would have to ignore anything beyond the hours at that point which might confuse the user.

fwolter replied:
Sounds tricky. But sounds tricky for the user, too. I won't insist on changing it.

ecdye replied:
I'm inclined to leave it as it is a lot of extra work. If I'm feeling adventurous at a later date I may try to implement this

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/OmnilinkDiscoveryService.java:
```diff
@@ -0,0 +1,517 @@
+            discoverHumiditySensors();
+            discoverLocks();
+        } catch (OmniInvalidResponseException | OmniUnknownMessageTypeException | BridgeOfflineException e) {
+            logger.debug("Received error during discovery", e);
```
Nope. I missed that one. I thought I got them all.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:
```diff
@@ -0,0 +1,678 @@
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            }
+        } catch (Exception e) {
```
Just a generic exception, it is because of upstream, there is no Specific exception defined. It is thrown when the controller is not accepting new connections

fwolter replied:
If I see correctly the upstream methods are declared with `throws OmniInvalidResponseException, OmniUnknownMessageTypeException, BridgeOfflineException`. You could catch these in a multi-catch.

fwolter replied:
I see, the constructor of  `Connection` in the library is delcared with `throws Exception, IOException, UnknownHostException`. This is a bit suboptimal, but not the scope of this review. You can leave it as is.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmnilinkHandlerFactory.java:
```diff
@@ -0,0 +1,145 @@
+            if (discovery != null) {
+                discovery.unregister();
+            }
+        }
```
Did I implement it correctly?

fwolter replied:
Looks good!

fwolter commented on bundles/org.openhab.binding.omnilink/pom.xml:
```diff
@@ -0,0 +1,26 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.0.0-SNAPSHOT</version>
```
To make the binding compile, you need to rebase your branch and set the version to the upcoming openHAB version: `3.1.0-SNAPSHOT`.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

fwolter reviewed (approved):
LGTM

digitaldan commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmnilinkHandlerFactory.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
```
@ecdye where is the controller/bridge thing returned here ?  I'm not sure how this is working for you as you cannot create a bridge/controller  thing.

digitaldan commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmnilinkHandlerFactory.java:
```diff
@@ -0,0 +1,100 @@
+            return new LockHandler(thing);
+        } else if (thingTypeUID.equals(THING_TYPE_CONSOLE)) {
+            return new ConsoleHandler(thing);
+        } else {
```
```suggestion
        } else if (thingTypeUID.equals(THING_TYPE_BRIDGE)) {
            return new OmnilinkBridgeHandler((Bridge) thing);
        } else {
```

digitaldan commented:
Once i added the bridge to be created by the HandlerFactory, I am getting this error:

```
2020-12-27 10:32:55.215 [WARN ] [hing.binding.BaseThingHandlerFactory] - Could not register service for class=org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService
java.lang.NoSuchMethodException: org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService.<init>()
```

I'll test this out some more, thanks @ecdye for the work so far.

ecdye commented:
I'll work on that issue, I haven't really tried to test the discovery too much so I'll work on it.

digitaldan commented:
Curious how the binding is working for you without the Factory returning a bridge?

ecdye commented:
> Curious how the binding is working for you without the Factory returning a bridge?

I haven't looked too closely at it, but yeah its not working.

ecdye commented:
@fwolter I'm not sure how to fix it but since I changed the Discovery Service implementation I can no longer scan to discover new items in the openHAB UI any ideas on how to fix this?

digitaldan commented:
@ecdye its hard for others to help if you force push every change as we can not see the change history, so it's not clear what the discovery looked like before and after.  I'm guessing you are relying on the bridge handler now being passed in by the framework, have you added logging to make sure this is happening, can you narrow down if anything is being called in that class?

digitaldan commented:
Also i noticed this is in the logs when i start the binding:
```
- Could not register service for class=org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService
java.lang.NoSuchMethodException: org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService.<init>()
```

ecdye commented:
I understand this, but what I am referring to is @fwolter directed me to change the Discovery Service implementation to use `ThingHandlerService` but since I implemented that I can no longer find the automatic discovery option in the openHAB UI.

Before the service was using a componenet annotation and was working properly. I simply can't find any documentation about this issue as the documentation provided by @fwolter does not indicate that anything else needs to be done for this to work.

digitaldan commented:
>but since I implemented that I can no longer find the automatic discovery option in the openHAB UI.

Did you see the error message i posted above  ^^,  there is an exception being thrown when it tries to register that class, its telling you  
`java.lang.NoSuchMethodException: org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService.<init>()` ,  
this means the class does not have an empty (no params) constructor, it needs a `public OmnilinkDiscoveryService(){...}` constructor.

ecdye commented:
> > but since I implemented that I can no longer find the automatic discovery option in the openHAB UI.
> 
> Did you see the error message i posted above ^^, there is an exception being thrown when it tries to register that class, its telling you
> `java.lang.NoSuchMethodException: org.openhab.binding.omnilink.internal.discovery.OmnilinkDiscoveryService.<init>()` ,
> this means the class does not have an empty (no params) constructor, it needs a `public OmnilinkDiscoveryService(){...}` constructor.

Apparently I cannot read, testing now.

randye007 commented:
@ecdye - Thanks so much for your work on this binding port! I can confirm that I can successfully discover my OMNI items and that so far all functionality is working well. Will let you know if I run into any issues.

ecdye commented:
@J-N-K Any chance you could help get this in before the next release?

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/LockHandler.java:
```diff
@@ -0,0 +1,120 @@
+                    sendOmnilinkCommand(OnOffType.OFF.equals(command) ? OmniLinkCmd.CMD_UNLOCK_DOOR.getNumber()
+                            : OmniLinkCmd.CMD_LOCK_DOOR.getNumber(), 0, thingID);
+                } else {
+                    logger.warn("Invalid command {}, must be OnOffType", command);
```
Just ignore unsupported commands (if you want to log, log at TRACE or DEBUG)

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmniLinkCmd.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    public static @Nullable OmniLinkCmd getCommand(String name) {
+        for (OmniLinkCmd command : OmniLinkCmd.values()) {
```
Why don't you use an `Optional` here? ou do that in other enums.

ecdye replied:
I did not write this code, I simply left it how it was as it was still working just fine. I try to operate on the principle of don't fix what isn't broken.

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:
```diff
@@ -0,0 +1,688 @@
+                        logger.debug("Could not send Set Time command to OmniLink Controller: {}", e.getMessage());
+                    }
+                } else {
+                    logger.warn("Invalid command: {}, must be DateTimeType", command);
```
see above

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:
```diff
@@ -0,0 +1,688 @@
+             * player.
+             */
+            audioPlayer = reqSystemFeatures().getFeatures().stream()
+                    .map(featureCode -> AudioPlayer.getAudioPlayerForFeatureCode(featureCode))
+                    .filter(Optional::isPresent).findFirst().orElse(Optional.empty());
```
```suggestion
                    .flatMap(AudioPlayer::getAudioPlayerForFeatureCode)
                    .filter(Objects::nonNull).findFirst();
```

ecdye replied:
This does not work 
``` 
[ERROR] /Users/ethandye/Developer/openhab-addons/bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:[237,70] The method flatMap(java.util.function.Function<? super java.lang.Integer,? extends java.util.stream.Stream<? extends R>>) in the type java.util.stream.Stream<java.lang.Integer> is not applicable for the arguments (AudioPlayer::getAudioPlayerForFeatureCode)
[ERROR] /Users/ethandye/Developer/openhab-addons/bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:[237,78] The type of getAudioPlayerForFeatureCode(int) from the type org.openhab.binding.omnilink.internal.AudioPlayer is java.util.Optional<org.openhab.binding.omnilink.internal.AudioPlayer>, this is incompatible with the descriptor's return type: java.util.stream.Stream<? extends R>
```

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/TemperatureFormat.java:
```diff
@@ -0,0 +1,91 @@
+
+    private final int formatNumber;
+
+    TemperatureFormat(int formatNumber) {
```
```suggestion
    private TemperatureFormat(int formatNumber) {
```

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/TemperatureFormat.java:
```diff
@@ -0,0 +1,91 @@
+     * @param omniNumber Number to convert
+     * @return Number converted to appropriate format.
+     */
+    abstract float omniToFormat(int omniNumber);
```
```suggestion
    public abstract float omniToFormat(int omniNumber);
```

also other methods

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/ThermostatHandler.java:
```diff
@@ -0,0 +1,241 @@
+                        TemperatureFormat.FAHRENHEIT.formatToOmni(((DecimalType) command).intValue()), thingID);
+                break;
+            default:
+                logger.warn("Unknown channel for Thermostat thing: {}", channelUID.getAsString());
```
```suggestion
                logger.warn("Unknown channel for Thermostat thing: {}", channelUID);
```

`ChannelUID` properly implements `toString`

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/units/FlagHandler.java:
```diff
@@ -0,0 +1,79 @@
+    public void updateChannels(ExtendedUnitStatus status) {
+        logger.debug("updateChannels called for Flag status: {}", status);
+        updateState(CHANNEL_FLAG_VALUE, DecimalType.valueOf(Integer.toString(status.getStatus())));
+        updateState(CHANNEL_FLAG_SWITCH, status.getStatus() == 0 ? OnOffType.OFF : OnOffType.ON);
```
`OnOffType.from`

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/units/UpbRoomHandler.java:
```diff
@@ -0,0 +1,183 @@
+        int unitStatus = status.getStatus();
+
+        updateState(CHANNEL_ROOM_STATE, new DecimalType(unitStatus));
+        updateState(CHANNEL_ROOM_SWITCH, Integer.valueOf(unitStatus).equals(1) ? OnOffType.ON : OnOffType.OFF);
```
```suggestion
        updateState(CHANNEL_ROOM_SWITCH, OnOffType.from(unitStatus == 1);
```

J-N-K reviewed (changes requested):
Thanks. I have left some comments inside. Please make sure to go through all occurences of the points I mentioned (regarding REFRESH handlig, `channelLinked` handling, logging, datatypes - DecimalType vs. QuantityType).

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/discovery/OmnilinkDiscoveryService.java:
```diff
@@ -0,0 +1,545 @@
+     * Discovers OmniLink audio zones
+     */
+    private void discoverAudioZones() {
+        final OmnilinkBridgeHandler handler = bridgeHandler;
```
Some of them it might be possible to, but because there are different supported properties and the way the Omni processes the requests, it would be difficult to do. As such, I would prefer to leave it for now.

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractAreaHandler.java:
```diff
@@ -0,0 +1,232 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("handleCommand: {}, command: {}", channelUID, command);
```
I'm not sure what you mean, what would that do?

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/ConsoleHandler.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
```
From my understanding it is because the OmniPro does not allow requesting the current state of these commands and therefore it cannot be requested.

ecdye replied:
@J-N-K I think I understand what you mean, however it is hard for me to tell if it is working. Does this look right?

ecdye commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/HumiditySensorHandler.java:
```diff
@@ -0,0 +1,144 @@
+    public void updateChannels(ExtendedAuxSensorStatus status) {
+        logger.debug("updateChannels called for Humidity Sensor status: {}", status);
+        updateState(CHANNEL_AUX_HUMIDITY,
+                new DecimalType(TemperatureFormat.FAHRENHEIT.omniToFormat(status.getTemperature())));
```
Because it already used DecimalType in the previous version of this binding and I did not think to change it, unless you think it is important, I would prefer to leave it as is.

digitaldan replied:
using a quantity type is simple, `new QuantityType<>(temp, temperatureUnit);` where temp is your float value and temperatureUnit is either `org.openhab.core.library.unit.ImperialUnits.FAHRENHEIT` or `org.openhab.core.library.unit.SIUnits.CELSIUS` depending on what the omni is reporting the value in

ecdye replied:
I'll work on implementing this, what do you use for humidity with QuantityType

ecdye replied:
@J-N-K Does that look better?

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/OmnilinkBridgeHandler.java:
```diff
@@ -612,6 +617,7 @@ private void updateBridgeProperties() {
     @Override
     public void initialize() {
         scheduleReconnectJob();
+        updateChannels();
```
Is this a good idea? What if the connection is not present? Wouldn't it make more sense to call `updateChannels` immediately after a successful connection?

ecdye replied:
Yes I thought about that and have changed it in a uncommited update that I am working on.

ecdye commented:
@J-N-K I have updated the files to all use REFRESH would you take a look and see what you think?

ecdye commented:
@J-N-K Any chance for a quick finish up of this PR?

J-N-K commented on bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractAreaHandler.java:
```diff
@@ -0,0 +1,245 @@
+        logger.debug("handleCommand: {}, command: {}", channelUID, command);
+
+        if (command instanceof RefreshType) {
+            Optional<ExtendedAreaStatus> status = retrieveStatus();
+            if (status.isPresent()) {
+                updateChannels(status.get());
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        "Received null staus update!");
+            }
```
```suggestion
            retrieveStatus().ifPresentOrElse(this::updateChannels, 
                () -> updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
                        "Received null staus update!"));
```
Also below

J-N-K reviewed (changes requested):
Last comment from my side.

ecdye commented:
@J-N-K Done. Can you approve and merge?

J-N-K commented:
Just waiting for Jenkins.

J-N-K commented:
What changed with your last push? If you force push the changes can‘t be shown.

ecdye commented:
Hmm, nothing changed... I updated for upstream change to allow PR to build correctly as @fwolter described.

J-N-K reviewed (approved):
Ok

boc-tothefuture commented:
As it appears we are close to merging this I want to thank @ecdye for pushing this forward as a PR and @J-N-K for his devotion to ensuring there is a maintainable code base that contains best practices. 

👏

ecdye commented:
All tests are good. I think we are ready.
